doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+# <Insert Name of Feature>
+
+## Summary
+
+A description of the correctness properties Materialize provides and the terms and concepts we use to describe them.
+
+## Goals
+
+The main goal is to lay the foundation for ongoing work to clearly articulate correctness properties of Materialize.
+This includes the introduction of concepts that underlie correctess, the components that will enforce them, and how they can communicate clearly.
+
+## Non-Goals
+
+It is not a goal to prescribe fine detail in how each of the components should behave, or which should have certain responsibilities.
+It is not a goal to prescribe which behavior Materialize should support, but only how to speak about a "correct" subset of it.
+It is not a goal to advocate (much) for DEFINITE over NON-VOLATILE.
+
+## Description
+
+Materialize means to provide several correctness guarantees, which manifest as various consistency and durability properties.
+At the heart of this is the idea that the ordering of events can be explicitly resolved and recorded, after which queries and views over the data become a matter of correct evaluation on completed subsets of time.
+The guarantees of Materialize derive from the durability of these recorded data, and the determinism of views over them.
+
+The primary term introduced here is "definite" to modify a collection of data.
```

Comment:
With respect to what observer does a collection need to present the same data at the same logical times to be definite? The choice will affect what kind of sources can be admitted into the system as definite.

Here are the options I see from weakest to strongest. A collection can be definite with respect to:

1. a single execution of a materialize instance (subsequent executions might present different data)
2. multiple executions of the same materialize instance (the instance gets to keep state around)
3. our singleton deployment of Materialize Platform (we get to keep state and share it among tenants)
4. any execution of any instance, regardless of its state

Option 1 isn't very useful, so let's focus on the rest. There is an important difference between 2/3 and 4. With options 2/3 we can take a collection that has no notion of offsets/timestamps (extreme example, reading `/dev/urandom`), and persist it with our invented timestamps, turning it into a definite collection. But with option 4 there is no way to take such a collection with no inherent notion of order and convert it into an ordered one because to do that we need local state, but 4 doesn't allow local state.

PubNub, and possibly other types of sources, falls into this category of "just a bunch of data", that I suppose we'd like to support and saying "our results are correct *within the Platform* or *within an instance*" is a strong enough statement.

Reply:
I think this has been cleared up in the new text. The only strong position taken here is that if there are names for definite collections, they should not be re-used with different meanings. One can have globally unique names, instance unique names, session unique names, any of these seem fine as long as they are not shared with folks for whom we cannot make the guarantee of definiteness.